components/odh-notebook-controller/e2e/notebook_creation_test.go

```diff
@@ -0,0 +1,229 @@
+			tc.testNotebookName, metav1.GetOptions{})
+
+		if err1 != nil {
+			if errors.IsNotFound(err1) {
+				return false, nil
+			} else {
+				log.Printf("Failed to get %s statefulset", tc.testNotebookName)
+				return false, err1
+			}
+		}
+		if notebookStatefulSet.Status.AvailableReplicas == 1 &&
+			notebookStatefulSet.Status.ReadyReplicas == 1 {
+			return true, nil
+		}
+		return false, nil
+	})
+	require.NoErrorf(t, err, "error validating notebook StatefulSet: %v", tc.testNotebookName)
+}
+
+func (tc *testContext) testNotebookOAuthSidecar(t *testing.T) {
+
+	nbPods, err := tc.kubeClient.CoreV1().Pods(tc.testNamespace).List(tc.ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("statefulset=%s", tc.testNotebookName)})
+
+	require.NoErrorf(t, err, "error retrieving Notebook pods")
+
+	for _, nbpod := range nbPods.Items {
+		require.Equalf(t, v1.PodRunning, nbpod.Status.Phase,
+			"Notebook pod %s are not in the Running phase", nbpod.Name)
+		for _, containerStatus := range nbpod.Status.ContainerStatuses {
+			if containerStatus.Name == "oauth-proxy" {
+				require.EqualValuesf(t, true, containerStatus.Ready,
+					"oauth-proxy container is not in Ready state ")
+			}
+		}
+	}
+}
+
+func (tc *testContext) testNotebookTraffic(t *testing.T) {
+	resp, err := tc.curlNotebookEndpoint()
+	require.NoErrorf(t, err, "error accessing Notebook Endpoint")
+	require.EqualValuesf(t, 200, resp.StatusCode, "Unexpected response from Notebook Endpoint")
+}
+
+func (tc *testContext) testNotebookCulling(t *testing.T) {
+	// Create Configmap with culling configuration
+	cullingConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "notebook-controller-culler-config",
+			Namespace: tc.testNamespace,
+		},
+		Data: map[string]string{
+			"ENABLE_CULLING":        "true",
+			"CULL_IDLE_TIME":        "2",
+			"IDLENESS_CHECK_PERIOD": "1",
+		},
+	}
+
+	_, err := tc.kubeClient.CoreV1().ConfigMaps(tc.testNamespace).Create(tc.ctx, cullingConfigMap,
+		metav1.CreateOptions{})
+	require.NoErrorf(t, err, "error creating configmap: notebook-controller-culler-config")
+	// Restart the deployment to get changes from configmap
+	controllerDeployment, err := tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Get(tc.ctx,
+		"notebook-controller-deployment", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	require.NoErrorf(t, tc.rolloutDeployment(controllerDeployment.ObjectMeta),
+		"error rolling out the deployment with culling configuration")
+
+	// Wait for server to shut down after 'CULL_IDLE_TIME' minutes(around 2.5 minutes)
+	time.Sleep(150 * time.Second)
+	// Verify that the notebook kernel has shutdown, and the notebook endpoint returns 503
+	resp, _ := tc.curlNotebookEndpoint()
+	require.EqualValuesf(t, 503, resp.StatusCode, "Unexpected response from Notebook Endpoint")
+	defer tc.revertCullingConfiguration(cullingConfigMap.ObjectMeta, controllerDeployment.ObjectMeta)
```

Comment:
By convention the `defer` clause is placed as soon as possible, not at the end:

```go
...
	// Restart the deployment to get changes from configmap
	controllerDeployment, err := tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Get(tc.ctx,
		"notebook-controller-deployment", metav1.GetOptions{})
	require.NoError(t, err)

	// Revert culling configuration when test is done
	defer tc.revertCullingConfiguration(cullingConfigMap.ObjectMeta, controllerDeployment.ObjectMeta)
...
```

For example -> https://github.com/dgraph-io/badger/blob/master/stream_writer_test.go#L179

Reply:
Done